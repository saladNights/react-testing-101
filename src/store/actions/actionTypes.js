export const SUBSCRIBE_SUCCESS = 'SUBSCRIBE_SUCCESS';
