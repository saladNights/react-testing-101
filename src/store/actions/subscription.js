import axios from 'axios';
import * as actionTypes from './actionTypes';

export const subscribeSuccess = () => {
	return {
		type: actionTypes.SUBSCRIBE_SUCCESS
	}
};

export const subscribe = (email) => {
	return dispatch => {
		return axios
			.post('https://react-testing-101.firebaseio.com/subscribers.json', { email })
			.then((response) => {
				const { status } = response;

				if (status === 200) dispatch(subscribeSuccess());
			});
	};
};
