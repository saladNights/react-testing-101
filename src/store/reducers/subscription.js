import * as actionTypes from '../actions/actionTypes';

const initialState = {
	success: false
};

export default (state=initialState, action) => {
	switch (action.type) {
		case actionTypes.SUBSCRIBE_SUCCESS:
			return {
				...state,
				...{ success: true }
			};
		default:
			return state;
	}
}
