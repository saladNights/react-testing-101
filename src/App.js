import React, { Component } from 'react';
import { connect } from 'react-redux';
import './App.css';

import { subscribe } from './store/actions/subscription'

class App extends Component {

  state = {
    email: '',
    formVisibility: false
  };

  changeEmail = (event) => {
    this.setState({ email: event.target.value});
  };

  toggleForm = () => {
    this.setState(prevState => ({ formVisibility: !prevState.formVisibility }))
  };

  submitForm = (e) => {
    e.preventDefault();
    const { onSubscribe } = this.props;
    const { email } = this.state;

    onSubscribe(email);
  };

  render() {
    const { email, formVisibility } = this.state;
    const { success } = this.props;

    return (
      <div className="App">
        <div className="form-wrapper">
          {success
            ? <span data-testid="success-msg" className="success-msg">Спасибо за подписку</span>
            : !formVisibility
              ? (<div>
                  <button className='btn_primary subscribe-btn' onClick={this.toggleForm} data-testid="subscribe">Подписаться на рассылку</button>
                </div>)
              : (<form data-testid="subscribe-form" className='subscribe-form'>
                  <input data-testid="subscribe-input" type="email" placeholder="Введите e-mail" value={email} onChange={this.changeEmail}/>
                  <button data-testid="subscribe-submit" className='btn_primary' onClick={this.submitForm}>Отправить</button>
                  <button className='btn_default' onClick={this.toggleForm}>Отмена</button>
                </form>)
          }
        </div>
      </div>
    );
  }
}

const mapStateToProps = state =>{
  const { success } = state.subscription;

  return { success };
};

const mapDispatchToProps = dispatch => {
  return {
    onSubscribe: (email) => dispatch(subscribe(email))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
