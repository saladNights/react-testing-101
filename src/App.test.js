import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';

import App from './App';
import { findByTestAttr, storeFactory } from './test/testUtils';
import moxios from "moxios";
import {subscribe} from "./store/actions/subscription";

describe('Component', () => {
  let wrapper;
  let store;
  const initialState = {
  	subscription: {
  		success: false
  	}
  };
	const props = {
		onSubscribe: jest.fn(),
		success: false
	};

  beforeEach(() => {
    wrapper = shallow(<App.WrappedComponent {...props} />);
  });

	it('renders without crashing', () => {
		expect(wrapper.exists()).toBe(true);
	});

	it("matches the snapshot", () => {
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it('show form', () => {
		const subscribeBtn = findByTestAttr(wrapper, 'subscribe');

		subscribeBtn.simulate('click');

		const form = findByTestAttr(wrapper, 'subscribe-form');

		expect(form.length).toBe(1);
	});

	// Тестирование реализации: bad example
	it('show form', () => {
		const subscribeBtn = wrapper.find('.subscribe-btn');

		subscribeBtn.simulate('click');

		const form = wrapper.find('.subscribe-form');

		expect(form.length).toBe(1);
	});

	it('should call subscribe with email', () => {
		const subscribeBtn = findByTestAttr(wrapper, 'subscribe');

		subscribeBtn.simulate('click');

		const input = findByTestAttr(wrapper, 'subscribe-input');

		input.simulate('change', {target: {value: 'blah@gmail.com'}});

		const submit = findByTestAttr(wrapper, 'subscribe-submit');
		const fakeEvent = { preventDefault: () => {} };

		submit.simulate('click', fakeEvent);

		expect(props.onSubscribe).toHaveBeenCalledWith('blah@gmail.com');
	});

	it('should show success message', async () => {
		store = storeFactory(initialState);
		wrapper = mount(<Provider store={store}><App {...props} /></Provider>);
		const subscribeBtn = findByTestAttr(wrapper, 'subscribe');

		moxios.install();

		subscribeBtn.simulate('click');

		const input = findByTestAttr(wrapper, 'subscribe-input');

		input.simulate('change', {target: {value: 'blah@gmail.com'}});

		const submit = findByTestAttr(wrapper, 'subscribe-submit');
		const fakeEvent = { preventDefault: () => {} };

		submit.simulate('click', fakeEvent);

		moxios.wait(() => {
			const request = moxios.requests.mostRecent();
			request.respondWith({
				status: 200
			});
		});

		await store.dispatch(subscribe());

		wrapper = wrapper.update();

		const successMessage = findByTestAttr(wrapper, 'success-msg');

		expect(successMessage.length).toBe(1);
		moxios.uninstall();
	});
});
