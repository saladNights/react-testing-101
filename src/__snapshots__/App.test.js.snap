// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches the snapshot 1`] = `
"<div className=\\"App\\">
  <div className=\\"form-wrapper\\">
    <div>
      <button className=\\"btn_primary subscribe-btn\\" onClick={[Function]} data-testid=\\"subscribe\\">
        Подписаться на рассылку
      </button>
    </div>
  </div>
</div>"
`;
